enarx-keep-sev-shim/src/hostcall.rs

```diff
@@ -0,0 +1,231 @@
+        let count = bytes.len();
+        assert!(count <= Block::buf_capacity());
+
+        let buf_address = Address::from(self.0.buf.as_mut_ptr());
+        let shim_virt_address = ShimVirtAddr::try_from(buf_address).map_err(|_| libc::EFAULT)?;
+
+        let request = Request::new(
+            Register::from_raw(libc::SYS_write as usize),
+            &[
+                fd.into(),
+                ShimPhysAddr::from(shim_virt_address).into(),
+                count.into(),
+            ],
+        );
+
+        self.0.msg.req = request;
+        self.0.buf[..count].copy_from_slice(&bytes[..count]);
+        self.hostcall().map(|r| r[0].raw() as _)
+    }
+
+    /// Exit the shim with a `status` code
+    ///
+    /// # Panics
+    ///
+    /// Panics, if the shim resumes to run.
+    #[inline(always)]
+    pub fn exit_group(&mut self, status: u32) -> ! {
+        unsafe {
+            let request = Request::new(
+                Register::from_raw(libc::SYS_exit_group as usize),
+                &[(status as usize).into()],
+            );
+            self.0.msg.req = request;
+
+            let _ = self.hostcall();
+
+            unreachable!()
+        }
+    }
+}
+
+impl<'a> Drop for HostCall<'a> {
+    fn drop(&mut self) {
+        // Put back in the virtual address in the global variable
+        let prev = unsafe { HOSTCALL_VIRT_ADDR.swap(self.0 as *mut _ as _, Ordering::Release) };
+        assert_eq!(prev, 0);
+    }
+}
+
+/// Write `bytes` to a host file descriptor `fd`
+#[inline(always)]
+pub fn shim_write(fd: HostFd, bytes: &[u8]) -> Result<libc::c_int, libc::c_int> {
+    let fd = usize::try_from(fd.as_raw_fd()).map_err(|_| libc::EBADF)?;
+    for c in bytes.chunks(Block::buf_capacity()) {
+        let chunk_len = c.len();
+        let mut to_write = chunk_len;
+
+        let mut host_call = HostCall::try_lock().ok_or(libc::EIO)?;
+
+        loop {
+            let written = unsafe {
+                host_call.write(
+                    fd,
+                    &c[(chunk_len.checked_sub(to_write).ok_or(libc::EFAULT)?)..],
+                )?
+            };
+            // be careful with `written` as it is untrusted
+            to_write = to_write
+                .checked_sub(usize::try_from(written).map_err(|_| libc::EIO)?)
+                .ok_or(libc::EIO)?;
+            if to_write == 0 {
+                break;
+            }
+        }
+    }
+
+    Ok(bytes.len().try_into().map_err(|_| libc::EIO)?)
+}
+
+/// Exit the shim with a `status` code
+///
+/// Reverts to a triple fault, which causes a `#VMEXIT` and a KVM shutdown,
+/// if it cannot talk to the host.
+pub fn shim_exit(status: u32) -> ! {
+    if let Some(mut host_call) = HostCall::try_lock() {
+        host_call.exit_group(status)
+    }
+
+    // provoke triple fault, causing a VM shutdown
+    unsafe { _enarx_asm_triple_fault() };
+    // in case the triple fault did not cause a shutdown
+    hlt_loop()
```

Comment:
Just as a comparison, in SGX we essentially do a host call in an infinite loop.

Reply:
In SEV-SNP, a VM `shutdown`, such as caused by the triple fault, cannot be resumed by the hypervisor, so I think we are safe here :-)